internal/provider/resource_application_test.go

```diff
@@ -268,6 +277,92 @@ func TestAcc_ResourceApplication_UpgradeProvider(t *testing.T) {
 	})
 }
 
+func TestAcc_ResourceApplication_EndpointBindings(t *testing.T) {
+	if testingCloud != LXDCloudTesting {
+		t.Skip(t.Name() + " only runs with LXD")
+	}
+	modelName := acctest.RandomWithPrefix("tf-test-application-bindings")
+	appName := "test-app"
+
+	managementSpace, publicSpace, cleanUp := setupModelAndSpaces(t, modelName)
+	defer cleanUp()
+
+	constraints := "arch=amd64 spaces=" + managementSpace + "," + publicSpace
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: frameworkProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				// test creating a single application with default endpoint bound to management space, and ubuntu endpoint bound to public space
+				PreConfig: func() {},
+				Config:    testAccResourceApplicationEndpointBindings(modelName, appName, constraints, map[string]string{"": managementSpace, "ubuntu": publicSpace}),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("juju_application.this", "model", modelName),
+					resource.TestCheckResourceAttr("juju_application.this", "endpoint_bindings.#", "2"),
+					resource.TestCheckTypeSetElemNestedAttrs("juju_application.this", "endpoint_bindings.*", map[string]string{"endpoint": "", "space": managementSpace}),
+					resource.TestCheckTypeSetElemNestedAttrs("juju_application.this", "endpoint_bindings.*", map[string]string{"endpoint": "ubuntu", "space": publicSpace}),
+					testCheckEndpointsAreSetToCorrectSpace(modelName, appName, managementSpace, map[string]string{"": managementSpace, "ubuntu": publicSpace}),
+				),
+			},
+		},
```

Comment:
Considering imports, I'm in a bind (pun intended).

Basically, in `computeEndpointBindingsWithoutDefaultValues`, I compare values from the state and the application's actual values to output only the necessary endpoints bindings to save in the state.

Necessary endpoints = either defined in previous state or different from default space
This works fine when updating, but not when importing.

This is because when importing the state is null.

here's an example of plan where it would not work:

consider the application top-ubuntu with endpoints bindings set as `"": "public"`

And this plan:
```tf
resource "juju_application" "ubuntu" {
  name  = "top-ubuntu"
  model = data.juju_model.resource.name

  constraints = "arch=amd64 spaces=management,public"
  charm {
    name    = "jameinel-ubuntu-lite"
    revision = 10
  }

  endpoint_bindings = [
    {
      space = "public"
    }
  ]
}

```

The space will not be stored in state since it's the same as defined in the application.
And I don't really have a way to differentiate what's default and what's provided by the user, because I don't have access to the plan on a read request.



TL;DR:
Can't test for import with endpoint bindings for the same reasons it's not tested for config. --> can't differentiate on import what's different between resource's defaults and user specified values that are same as defaults.
Same behavior as config

Reply:
We need to think of this for real life. I'm sure someone will try it. 

We do handle it for config, as there is code to differentiate between default and changed application config value. It might not be tested.

Reply:
Ok, I used the model's config option `default-space` to be able to compute what's a default value or not. I was reticent to use it, but I don't think there's another way to know if a space is default or not.